src/utils/cache.h

```diff
@@ -186,121 +207,99 @@ class LruCache {
+  void EvictItem() REQUIRES(mutex_) {
     --size_;
-
-    // Remove from LRU list.
-    if (lru_head_ == iter) {
-      lru_head_ = iter->next_in_queue;
-    } else {
-      iter->prev_in_queue->next_in_queue = iter->next_in_queue;
+    uint64_t key = insertion_order_.front();
+    insertion_order_.pop_front();
+    size_t idx = key % hash_.size();
+    while (true) {
+      if (hash_[idx].metadata == 1 && hash_[idx].key == key) {
+        break;
+      }
+      ++idx;
+      if (idx >= hash_.size()) idx -= hash_.size();
     }
-    if (lru_tail_ == iter) {
-      lru_tail_ = iter->prev_in_queue;
+    if (hash_[idx].pins == 0) {
+      --allocated_;
+      hash_[idx].value.reset();
+      hash_[idx].metadata = 0;
     } else {
-      iter->next_in_queue->prev_in_queue = iter->prev_in_queue;
+      evicted_.emplace_back(hash_[idx].key, std::move(hash_[idx].value));
+      evicted_.back().pins = hash_[idx].pins;
+      hash_[idx].pins = 0;
+      hash_[idx].metadata = 0;
     }
-
-    // Destroy or move into evicted list depending on whether it's pinned.
-    Item** cur = &hash_[hasher_(iter->key) % hash_.size()];
-    for (Item* el = *cur; el; el = el->next_in_hash) {
-      if (el == iter) {
-        *cur = el->next_in_hash;
-        if (el->pins == 0) {
-          --allocated_;
-          delete el;
-        } else {
-          el->next_in_hash = evicted_head_;
-          evicted_head_ = el;
-        }
-        return;
+    size_t next = idx + 1;
+    if (next >= hash_.size()) next -= hash_.size();
+    while (true) {
+      if (hash_[next].metadata == 0) {
+        break;
       }
-      cur = &el->next_in_hash;
-    }
-
-    assert(false);
-  }
-
-  void ShrinkToCapacity(int capacity) REQUIRES(mutex_) {
-    if (capacity < 0) capacity = 0;
-    while (lru_tail_ && size_ > capacity) {
-      EvictItem(lru_tail_);
+      size_t target = hash_[next].key % hash_.size();
+      if (!InRange(target, idx + 1, next)) {
+        std::swap(hash_[next], hash_[idx]);
+        idx = next;
+      }
+      ++next;
+      if (next >= hash_.size()) next -= hash_.size();
     }
   }
 
-  void BringToFront(Item* iter) REQUIRES(mutex_) {
-    if (lru_head_ == iter) {
-      return;
+  bool InRange(size_t target, size_t start, size_t end) {
+    if (start <= end) {
+      return target >= start && target <= end;
     } else {
-      iter->prev_in_queue->next_in_queue = iter->next_in_queue;
+      return target >= start || target <= end;
     }
-    if (lru_tail_ == iter) {
-      lru_tail_ = iter->prev_in_queue;
-    } else {
-      iter->next_in_queue->prev_in_queue = iter->prev_in_queue;
-    }
-
-    InsertIntoLru(iter);
   }
 
-  void InsertIntoLru(Item* iter) REQUIRES(mutex_) {
-    iter->next_in_queue = lru_head_;
-    iter->prev_in_queue = nullptr;
-
-    if (lru_head_) {
-      lru_head_->prev_in_queue = iter;
-    }
-    lru_head_ = iter;
-    if (lru_tail_ == nullptr) {
-      lru_tail_ = iter;
+  void ShrinkToCapacity(int capacity) REQUIRES(mutex_) {
```

Comment:
Maybe `EvictToCapacity`?
(I know it was in the original code, but now as I read it, Evict sounds clearer)

Reply:
Done.